text/0001-package-structure.md

```diff
@@ -0,0 +1,135 @@
+* Consistent tooling across the project
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+It is proposed that we start with a mono-repo structure consisting of multiple packages,
+each with their own set of dependencies.
+The tooling and CI will be shared across all projects.
+There will be a `packages` directory at the root of the repo, with each package in its own subdirectory.
+An example has been implemented in [CMIP-REF/cmip-ref#1](https://github.com/CMIP-REF/cmip-ref/pull/0001),
+which contains a `ref-core` package and an example `ref-metrics-example` package.
+
+Each package will have a `README.md` file that describes the purpose of the package and how to use it,
+its own `pyproject.toml` describing the package and it's dependencies, 
+and their own set of tests.
+The test suite for each package is run with only the dependencies required for that package.
+
+A common set of integration tests will also be run to ensure that the packages work together.
+
+The purpose of this structure is to split the project into smaller, more manageable parts
+and to differentiate between the "library" part of the project which is reusable
+and the "application" which builds ontop of the library + metrics packages + additional services.
+
+An additional benefit is that it will be easier to refactor code if it is all accessible in the same repository.
+
+## Packages
+### `ref-core`
+This package will be a library containing the core functionality of the REF.
+This package will be a dependency of all other packages as it will describe the interfaces that metrics providers must implement. 
+This allows us to keep the core functionality separate from the metrics providers.
+
+### `ref-metrics-*`
+Each metric provider will implement a separate package that implements the interfaces described by the `ref-core` package.
+These will very thin packages that wrap around the existing APIs of the benchmarking packages or use CMEC to execute.
+
+An example package `ref-metrics-example` has been implemented in [CMIP-REF/cmip-ref#5](https://github.com/CMIP-REF/cmip-ref/pull/5).
+See the [code](https://github.com/CMIP-REF/cmip-ref/tree/basic-interface/packages/ref-metrics-example) for more details.
+
+The purpose of these packages are to provide a Python-based hook to execute a metric and return the results.
+This will include (hopefully thin) wrappers around the metric providers existing APIs
+or build the required CMEC configuration files.
+It should require little to no change to the existing codebase benchmarking packages.
+
+The repository will set up a CI pipeline that will run the unit tests for each package.
+This includes fetching some sample data.
+
+These packages will maintain their own dependencies (python and otherwise) and packaged up as a docker image.
+
+The package will be named `ref-metrics-<provider>` where `<provider>` is the name of the provider.
+
+### `ref-api` (TBD)
```

Comment:
The way I understand it, there are at least two use cases that each need their own application.
1) An automated system running on an ESGF node that listens to the ESGF message queue announcing that new datasets have been published and then kicks off the required diagnostic runs and integrates/publishes the results.
1) Human users who want to run the tool themselves, this could be from the command line or from a Jupyter notebook

It looks like this proposes 1), but maybe it would be more convenient to start with 2) as having that is convenient for development. Or is that covered somewhere else already?

Reply:
I would add/tweak another one:
1) Production deployment
2) Modelling center where they are ingesting their pre-published data. Probably by running a CLI command. This requires a bulk of the services needed for a production deployment if they want to track results.
3) Running metrics directly via a common interface without any tracking of results

I'm assuming that modelling centers want some form of tracking, but that is only a guess so we should validate that. The tracking requires a database which in likely just sqlite which doesn't require any additional services.

There is a lot of overlap of 1 and 2, but 3 is probably how benchmarking package maintainers will develop their packages. Each of the metrics packages should be able to be run directly in a notebook as well if you don't require the complexity that comes from the "compute engine".